python/paddle/distributed/auto_parallel/api.py

```diff
@@ -1566,3 +1565,182 @@ def unshard_dtensor(dist_tensor):
+    """
+    ShardDataloader converts a dataloader to a new dataloader which provided two capabilities:
+    1. split dataloader by shard_dim to do data parallel.
+    2. reshard the output of dataloader to distributed tensor.
+    if has_dataset_splitted is True, just need to do reshard.
+
+    Args:
+        dataloader (paddle.io.DataLoader): The dataloader to be sharded.
+        meshes (Union(list, tuple)): The mesh list of the dataloader. Identify which mesh the input is on.
+        input_keys Union(list, tuple): The input keys of the outputs of dataloader if it is a dict.
+            The input corresponding to input_keys[i] is on meshes[i]. if outputs is a list, no need to pass it,
+            and the i'th input is on meshes[i].
+        shard_dims (Union[list, tuple, str, int]): The mesh dimension to shard the dataloader.
+            Users can specify the shard_dim of each mesh or specify a single shard_dim for all meshes.
+        has_dataset_splitted (bool): Whether the dataset has been splitted.
+    """
+
+    def __init__(
+        self,
+        dataloader: paddle.io.DataLoader,
+        meshes: List[dist.ProcessMesh],
+        input_keys: List[str] = None,
+        shard_dims: Union[list, tuple, str, int] = None,
+        has_dataset_splitted: bool = False,
+    ):
+        # do some check
+        if has_dataset_splitted is True and shard_dims is None:
+            raise ValueError(
+                "shard_dims must be set when has_dataset_splitted is True"
+            )
+        self._meshes = meshes
+        self._all_inputs_in_one_mesh = len(self._meshes) == 1
+        self._input_keys = input_keys
+        self._shard_dims = self.process_shard_dims(shard_dims)
+
+        process_id = dist.get_rank()
+        mesh_index = self.current_process_in_which_mesh(process_id)
+        if mesh_index == -1:
+            dp_rank = 0
+            dp_world_size = self._meshes[0].get_dim_size(self._shard_dims[0])
+        else:
+            dp_rank = self._meshes[mesh_index].get_rank_by_dim(
+                self._shard_dims[mesh_index], process_id
+            )
+            dp_world_size = self._meshes[mesh_index].get_dim_size(
+                self._shard_dims[mesh_index]
+            )
+
+        if has_dataset_splitted is True or shard_dims is None:
+            self._dataloader = dataloader
+            self.batch_size = dataloader.batch_sampler.batch_size
+        else:
+            self.batch_size = int(
+                dataloader.batch_sampler.batch_size / dp_world_size
+            )
+            self.batch_sampler = DistributedBatchSampler(
+                dataset=dataloader.dataset,
+                batch_size=self.batch_size,
+                num_replicas=dp_world_size,
+                rank=dp_rank,
+            )
+            self._dataloader = paddle.io.DataLoader(
+                dataset=dataloader.dataset, batch_sampler=self.batch_sampler
+            )
+
+    def process_shard_dims(self, shard_dims):
```

Comment:
-> _process_shard_dims, same for other non-public interfaces.

Reply:
Done